src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+// =====================
+
+/// The semantic split, splits the sub-shapes according to their relative position from `OutputPort`
+/// to `InputPort` and allows access to the three different groups of shapes:
+///   - shapes that are input side of the split;
+///   - shapes that are at the split;
+///   - shapes that are output side of the split.
+///
+/// Note that "at the split" also includes the shapes adjacent to the actual split because they
+/// need to be treated as if they were at the split location to avoid glitches at the shape
+/// boundaries.
+/// // FIXE: please, tell more about these glitches - I do not understand why they happen after this description
+///
+/// This allows us to apply special handling to these groups. This is required as a simple geometric
+/// split based on a line, will lead to double intersections with the edge. Thus we avoid the
+/// geometric intersections, for shapes away from the intersection point, and instead color them
+/// based on their position within the edge.
+///
+/// We have seven "slots" of shapes within the edge that can be ordered from output port to input
+/// port: `SideLine` `Corner`, `MainLine`/`Arrow`, `Corner2`, `SideLine2`, `Corner3`, `PortLine`.
+///
+/// Example: We need to split on the `SideLine2` and highlight the shapes closer to the
+/// output port. That means we need to do the geometric split on  `Corner2`, `SideLine2`, `Corner3`,
+/// which we can access via `split_shapes` and apply the highlighting to `SideLine` `Corner` and
+/// `MainLine`/`Arrow`, which we can access via `output_side_shapes`. The remaining shapes that must
+/// not be highlighted can be accessed via `input_side_shapes`.
+#[derive(Clone,Debug)]
+struct SemanticSplit {
+    /// Ids of the shapes in the order they appear in the edge. Shapes that fill the same "slot" in
+    /// the shape and must be handled together, are binned into a sub-vector. That can be the case
+    /// for shapes that are present in the back and the front of the shape.
+    ordered_part_ids : Vec<Vec<display::object::Id>>,
+    /// The index the shape where the edge split occurs in the `ordered_part_ids`.
+    split_index      : usize,
+}
+
+impl SemanticSplit {
+
+    fn new(edge_data:&EdgeModelData, split_shape:display::object::Id) -> Option<Self> {
+        let ordered_part_ids = Self::semantically_binned_edges(edge_data);
+
+        // Find the object id in our `ordered_part_ids`
+        let mut split_index  = None;
+        for (index, shape_ids) in ordered_part_ids.iter().enumerate() {
+            if shape_ids.contains(&split_shape) {
+                split_index = Some(index);
+                break
+            }
+        }
+        let split_index = split_index?;
+
+        Some(SemanticSplit {ordered_part_ids,split_index})
+    }
+
+    /// Return a ordered vector that contains the ids of the shapes in the order they appear in the
+    /// edge. Shapes that are to be handled as in the same place, are binned into a sub-vector.
```

Comment:
Are you sure you wanted to use the "binned" verb here? My diet tells it means that something is put in a bin, but this could also be my weak english

Reply:
They are put in a bin :) This is derived from [data binning/bucketing](https://en.wikipedia.org/wiki/Data_binning)